p2p: change `CInv::type` from `int` to `uint32_t`, fix UBSan warning

jonatack opened the pull request:
Fixes UBSan implicit-integer-sign-change issue per https://github.com/bitcoin/bitcoin/pull/19610#issuecomment-680686460.

Credit to Crypt-iQ for finding and reporting the issue and to vasild for the original review suggestion in https://github.com/bitcoin/bitcoin/pull/19590#pullrequestreview-455788826.

Closes #19678.

jonatack committed:
p2p: change CInv::type from int to uint32_t

fixes issue #19678 UBSan implicit-integer-sign-change

Credit to Eugene (Crypt-iQ) for finding and reporting the issue
and to Vasil Dimov (vasild) for the original suggestion

laanwj commented on src/protocol.h:
```diff
@@ -425,7 +425,7 @@ class CInv
     bool IsGenTxMsg()     const { return type == MSG_TX || type == MSG_WTX || type == MSG_WITNESS_TX; }
 
-    int type;
+    uint32_t type;
```
ACK on using a sized type here.
However, changing from signed to unsigned is, in principle, a P2P protocol change.
This needs to be carefully reviewed for potential unexpected by-effects.

MarcoFalke commented:
According to https://en.bitcoin.it/wiki/Protocol_documentation#Inventory_Vectors (section written in 2010 by "MagicalTux") this is unsigned. Though, our python tests also treat this as signed. https://btcinformation.org/en/developer-reference#inv doesn't say anything about signedness.

Would be good to at least update the python test framework and also add a p2p test.

jonatack committed:
test framework: serialize/deserialize inv type as unsigned int

jonatack commented:
> Would be good to at least update the python test framework and also add a p2p test.

Updated the python test framework and added a test. Was unsure what kind of test to add; this one tests the framework change.

jonatack commented:
> Would be good to at least update the python test framework and also add a p2p test.

Updated the python test framework and added a test for the framework change.

MarcoFalke commented on test/functional/p2p_invalid_messages.py:
```diff
@@ -125,6 +128,16 @@ def test_msgtype(self):
+        invalid_inv = msg_inv([CInv(t=-1, h=1)])  # inv type must be an unsigned integer
+        try:
+            self.nodes[0].add_p2p_connection(P2PInterface()).send_message(invalid_inv)
+        except struct.error as e:
```
We generally don't test the test framework. Sending an uint32_t{-1} and observing the node doesn't crash with sanitizers enabled should be a sufficient smoke test. Maybe add an assert_debug_log to observe how the inv will be formatted.

```
            LogPrint(BCLog::NET, "got inv: %s  %s peer=%d\n", inv.ToString(), fAlreadyHave ? "have" : "new", pfrom.GetId());

jonatack replied:
Thanks for the tip. The temporary framework test helped me see an issue with the framework update, will drop it and try your suggestion.

jonatack replied:
Thanks Marco. Trying some stuff to see how the CI reacts to it.

MarcoFalke commented on test/functional/p2p_invalid_messages.py:
```diff
@@ -125,6 +129,27 @@ def test_msgtype(self):
+
+        # Repeat, passing signed integer CInv type format to override default.
+        LE_signed_int = "<i"  # Python format chars for little-endian signed int
+        inv_with_signed_type = msg_inv([CInv(t=-MSG_TX, h=1, f=LE_signed_int)])
```
```suggestion
        inv_with_signed_type = msg_inv([CInv(t=0xffffffff, h=1)])
```

Serialized -1 as signed integer is identical to 0xffffffff serialized as unsigned integer, so might as well use the latter

jonatack replied:
That was breaking this method in messages.py::CInv with KeyError: 4294967295
```python
    def __repr__(self):
        return "CInv(type=%s hash=%064x)" \
            % (self.typemap[abs(self.type)], self.hash)
```

MarcoFalke replied:
Even when adding type `0xffffffff: "Error"`?

jonatack replied:
Travis seem happy. I have the following minor cleanups in mind but will wait for feedback before pushing.
```diff
test/functional/p2p_invalid_messages.py
@@ -142,12 +142,12 @@ class InvalidMessagesTest(BitcoinTestFramework):
         LE_signed_int = "<i"  # Python format chars for little-endian signed int
         inv_with_signed_type = msg_inv([CInv(t=-MSG_TX, h=1, f=LE_signed_int)])
-        with self.nodes[0].assert_debug_log(["got inv: 0xffffffff 00000"]):
+        with self.nodes[0].assert_debug_log(["got inv: 0xffffffff "]):
             self.nodes[0].add_p2p_connection(P2PInterface()).send_message(inv_with_signed_type)
 
         # Sanity check with valid unsigned inv type and same format override.
         inv_with_unsigned_type = msg_inv([CInv(t=MSG_WTX, h=1, f=LE_signed_int)])
-        with self.nodes[0].assert_debug_log(["got inv: wtx 0000"]):
+        with self.nodes[0].assert_debug_log(["got inv: wtx "]):

test/functional/test_framework/messages.py
@@ -251,7 +251,7 @@ class CInv:
     def __init__(self, t=0, h=0, f="<I"):
         self.type = t
         self.hash = h
-        self.inv_type_format = f  # default format is unsigned int
+        self.inv_type_format = f  # default is little endian unsigned int
```

jonatack replied:
> Even when adding type `0xffffffff: "Error"`?

Didn't think of it :D thanks, trying.

jonatack replied:
Great--that simplifies things.

jonatack replied:
Done, simplified, good idea. LMK if we should drop the first test.

practicalswift commented:
Concept ACK

Thanks for fixing UBSan fuzzing finds!

practicalswift commented:
@jonasnick Just to confirm: if the test changes we're dropped in this change set (keeping only the changes to `src/protocol.{cpp,h}`) then all functional tests would still pass, right?

MarcoFalke commented on test/functional/p2p_invalid_messages.py:
```diff
@@ -125,6 +128,22 @@ def test_msgtype(self):
+        inv_with_signed_type = msg_inv([CInv(t=-MSG_TX, h=1)])
+        try:
+            self.nodes[0].add_p2p_connection(P2PInterface()).build_message(inv_with_signed_type)
+        except struct.error as e:
```
imo could drop the test framework self-test

jonatack replied:
Done, but I think both of these tests only tested the framework change anyway.

jonatack commented:
> @jonasnick Just to confirm: if the test changes we're dropped in this change set (keeping only the changes to `src/protocol.{cpp,h}`) then all functional tests would still pass, right?

Right. I pushed the changes to `src/protocol.{h/cpp}` several times in other PR changesets (#19610, #19611) before dropping them, and the CI passes.

jnewbery commented:
> According to https://en.bitcoin.it/wiki/Protocol_documentation#Inventory_Vectors (section written in 2010 by "MagicalTux") this is unsigned. Though, our python tests also treat this as signed. https://btcinformation.org/en/developer-reference#inv doesn't say anything about signedness.

btcinformation.org does show this as uint32_t: https://btcinformation.org/en/developer-reference#data-messages. That was added by @harding to the bitcoin.org developer docs here: https://github.com/bitcoin-dot-org/Bitcoin.org/commit/1634212dd5550bb9b610341668b4b7b5fa3690a5#diff-693f5b3e9d88b97c899694c8d1656cbbR110

Note that even though comparing an int to a uint32_t causes a UBSan warning, it's not actually undefined behaviour: https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#:~:text=%2Dfsanitize%3Dimplicit%2Dinteger%2D,the%20new%20value%20is%20negative. However, I think it's still worth changing this, so that we're always comparing / xor'ing a uint32_t with a uint32_t, so concept ACK.

jonatack commented:
Great links @jnewbery, thanks.

> Note that even though comparing an int to a uint32_t causes a UBSan warning, it's not actually undefined behaviour: [clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#:~:text=%2Dfsanitize%3Dimplicit%2Dinteger%2D,the%20new%20value%20is%20negative](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#:~:text=%2Dfsanitize%3Dimplicit%2Dinteger%2D,the%20new%20value%20is%20negative).

"-fsanitize=implicit-integer-sign-change: Implicit conversion between integer types, if that changes the sign of the value...issues caught by this sanitizer are not undefined behavior, but are often unintentional."

laanwj commented on test/functional/test_framework/messages.py:
```diff
@@ -244,27 +244,28 @@ class CInv:
     def __repr__(self):
         return "CInv(type=%s hash=%064x)" \
-            % (self.typemap[self.type], self.hash)
+            % (self.typemap[abs(self.type)], self.hash)
```
I don't think this `abs(…)` is correct. There should be no reason to take an absolute value here, convert, say, `-1` to `1`.

jonatack replied:
It was to not allow shallow errors of type `KeyError` to hide deeper issues.

jonatack replied:
Dropped this line.

laanwj commented on test/functional/test_framework/messages.py:
```diff
@@ -244,27 +244,28 @@ class CInv:
-        5: "WTX",
+        MSG_CMPCT_BLOCK: "CompactBlock",
+        MSG_WTX: "WTX",
+        0xffffffff: "Error",
```
I'd prefer to report anything that is unkonwn as "Error" instead of specifically adding `0xffffffff`  here.

jonatack replied:
This was only needed for the test. I tried a few [different tests](https://github.com/bitcoin/bitcoin/commit/6c0f4aefe373bb89188e91464326dabf0f9f3397]. [This one](https://github.com/bitcoin/bitcoin/commit/1ab179b78d63c7e96af7dbad6caf3efb82afa01a) doesn't require adding entries to `typemap`. I don't think any of the tests cover the code change, only the test framework change.

jonatack replied:
I'm not sure what to do here. This was only needed for the test. I tried a few [different tests](https://github.com/bitcoin/bitcoin/commit/6c0f4aefe373bb89188e91464326dabf0f9f3397); the first one doesn't require adding entries to `typemap`. I don't think any of the tests cover the CInv::type code change, only the test framework change.

jonatack replied:
Dropped this line.

jonatack commented:
Thanks everyone for your feedback. I wrote a [few tests](https://github.com/bitcoin/bitcoin/commit/6c0f4aefe373bb89188e91464326dabf0f9f3397) but they all covered the change to the CInv class in the test framework, not the CInv change in `protocol.{h,cpp}`, so I've dropped them in the latest push.

laanwj commented:
Thanks for addressing the comments, the code changes look good to me now.

vasild reviewed (approved):
ACK 7984c39be

All the values we ever assign to `CInv::type` are non-negative within the reach of a signed 32 bit integer. `GetDataMsg` is `uint32_t` and in some places we `|=` a variable of type `uint32_t` into `CInv::type`:

https://github.com/bitcoin/bitcoin/blob/bab4cce1b0eedc1a51692aaf83ba54dd0a9d17e6/src/net_processing.cpp#L2717

So it makes sense to have `CInv::type` as `uint32_t`.

Actually serializing a signed integer in the way we do it is not portable, it just works because all platforms use [two's complement](https://en.wikipedia.org/wiki/Two%27s_complement), but that's not guaranteed.

laanwj commented:
> All the values we ever assign to CInv::type are non-negative within the reach of a signed 32 bit integer. GetDataMsg is uint32_t and in some places we |= a variable of type uint32_t into CInv::type

Bitwise arithmetic should definitely use unsigned integers, good point.

laanwj commented:
ACK 407175e0c2bc797599ebd9c0a1f2ec89ad7af136
